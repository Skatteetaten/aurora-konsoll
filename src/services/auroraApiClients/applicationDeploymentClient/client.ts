import GoboClient, { IGoboResult } from 'services/GoboClient';
import {
  REDEPLOY_WITH_CURRENT_VERSION_MUTATION,
  REDEPLOY_WITH_VERSION_MUTATION,
  REFRESH_APPLICATION_DEPLOYMENT_MUTATION,
  REFRESH_APPLICATION_DEPLOYMENTS_MUTATION
} from './mutation';
import {
  APPLICATION_DEPLOYMENT_DETAILS_QUERY,
  APPLICATIONS_QUERY,
  IApplicationDeploymentDetailsQuery,
  IApplicationsConnectionQuery,
  IUserAffiliationsQuery,
  USER_AFFILIATIONS_QUERY
} from './query';

export function formatName(user: string) {
  const names = user.split(', ');
  if (names.length !== 2) {
    return user;
  }
  return names[1] + ' ' + names[0];
}

export class ApplicationDeploymentClient {
  private client: GoboClient;

  constructor(client: GoboClient) {
    this.client = client;
  }

  public async redeployWithVersion(
    applicationDeploymentId: string,
    version: string
  ): Promise<IGoboResult<{ redeployWithVersion: boolean }> | undefined> {
    return await this.client.mutate<{ redeployWithVersion: boolean }>({
      mutation: REDEPLOY_WITH_VERSION_MUTATION,
      variables: {
        input: {
          applicationDeploymentId,
          version
        }
      }
    });
  }

  public async redeployWithCurrentVersion(
    applicationDeploymentId: string
  ): Promise<IGoboResult<{ redeployWithCurrentVersion: boolean }> | undefined> {
    return await this.client.mutate<{
      redeployWithCurrentVersion: boolean;
    }>({
      mutation: REDEPLOY_WITH_CURRENT_VERSION_MUTATION,
      variables: {
        input: {
          applicationDeploymentId
        }
      }
    });
  }

  public async refreshApplicationDeployment(
    applicationDeploymentId: string
  ): Promise<
    IGoboResult<{ refreshApplicationDeployment: string }> | undefined
  > {
    return await this.client.mutate<{
      refreshApplicationDeployment: string;
    }>({
      mutation: REFRESH_APPLICATION_DEPLOYMENT_MUTATION,
      variables: {
        input: {
          applicationDeploymentId
        }
      }
    });
  }

  public async refreshAffiliations(
    affiliations: string[]
  ): Promise<IGoboResult<{ affiliations: string[] }> | undefined> {
    return await this.client.mutate<{
      affiliations: string[];
    }>({
      mutation: REFRESH_APPLICATION_DEPLOYMENTS_MUTATION,
      variables: {
        input: {
          affiliations
        }
      }
    });
  }

  public async findApplicationDeploymentDetails(
    id: string
  ): Promise<IGoboResult<IApplicationDeploymentDetailsQuery> | undefined> {
    return await this.client.query<IApplicationDeploymentDetailsQuery>({
      query: APPLICATION_DEPLOYMENT_DETAILS_QUERY,
      variables: {
        id
      }
    });
  }

  public async findUserAndAffiliations(): Promise<
    IGoboResult<IUserAffiliationsQuery> | undefined
  > {
    return await this.client.query<IUserAffiliationsQuery>({
      query: USER_AFFILIATIONS_QUERY
    });
  }

  public async findAllApplicationDeployments(
    affiliations: string[]
  ): Promise<IGoboResult<IApplicationsConnectionQuery> | undefined> {
    return await this.client.query<IApplicationsConnectionQuery>({
      query: APPLICATIONS_QUERY,
      variables: {
        affiliations
      }
    });
  }
}

// ! Temp fix for template deployments with default version
// TODO: FIX
export function findDeployTagForTemplate(
  applicationName: string,
  deployTag: string
) {
  const templates = {
    'aurora-activemq-1.0.0': '2',
    'aurora-redis-1.0.0': '3.2.3',
    'aurora-wiremock-1.0.0': '1.3.0',
    redis: '3.2.3',
    wiremock: '1.3.0'
  };

  if (deployTag) {
    return deployTag;
  }

  return templates[applicationName] || deployTag;
}
