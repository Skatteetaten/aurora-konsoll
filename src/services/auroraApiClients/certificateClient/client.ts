import { ICertificateResult } from 'models/certificates';
import GoboClient from 'services/GoboClient';
import { CERTIFICATES_QUERY, ICertificate, ICertificatesQuery } from './query';

export class CertificateClient {
  public defaultCertificates: ICertificatesQuery = {
    certificates: {
      totalCount: 0,
      edges: []
    }
  };

  public client: GoboClient;

  constructor(client: GoboClient) {
    this.client = client;
  }
  public async getCertificates(): Promise<ICertificateResult> {
    const result = await this.client.query<ICertificatesQuery>({
      query: CERTIFICATES_QUERY
    });
    if (!result) {
      return {
        certificates: [],
        totalCount: 0
      };
    }

    return {
      certificates: this.normalizeScanStatus(result.data),
      totalCount: result.data.certificates.totalCount
    };
  }

  private normalizeScanStatus(data: ICertificatesQuery): ICertificate[] {
    if (!data) {
      return [];
    }

    return data.certificates.edges.map(edge => {
      const { cn, expiresDate, id, issuedDate, revokedDate } = edge.node;
      return {
        issuedDate,
        cn,
        expiresDate,
        id,
        revokedDate
      };
    });
  }
}
