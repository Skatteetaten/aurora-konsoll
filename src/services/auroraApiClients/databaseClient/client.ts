import GoboClient, { IDataAndErrors } from 'services/GoboClient';

import {
  ICreateDatabaseSchemaInput,
  IDeleteDatabaseSchemasResponse,
  IJdbcUser,
  IUpdateDatabaseSchemaInputWithCreatedBy
} from 'models/schemas';
import {
  CREATE_DATABASE_SCHEMA_MUTATION,
  DELETE_DATABASESCHEMAS_MUTATION,
  TEST_JDBC_CONNECTION_FOR_ID_MUTATION,
  TEST_JDBC_CONNECTION_FOR_JDBCUSER_MUTATION,
  UPDATE_DATABASESCHEMA_MUTATION
} from './mutation';
import {
  DATABASE_SCHEMAS_QUERY,
  IDatabaseSchemasQuery,
  DATABASE_INSTANCES_QUERY,
  IDatabaseInstancesQuery
} from './query';

export class DatabaseClient {
  private client: GoboClient;

  constructor(client: GoboClient) {
    this.client = client;
  }

  public async getSchemas(
    affiliations: string[]
  ): Promise<IDataAndErrors<IDatabaseSchemasQuery> | undefined> {
    return await this.client.query<IDatabaseSchemasQuery>({
      query: DATABASE_SCHEMAS_QUERY,
      variables: {
        affiliations
      }
    });
  }

  public async getInstances(
    affiliation: string
  ): Promise<IDataAndErrors<IDatabaseInstancesQuery> | undefined> {
    return await this.client.query<IDatabaseInstancesQuery>({
      query: DATABASE_INSTANCES_QUERY,
      variables: {
        affiliation
      }
    });
  }

  public async updateSchema(
    databaseSchema: IUpdateDatabaseSchemaInputWithCreatedBy
  ): Promise<
    IDataAndErrors<{ updateDatabaseSchema: { id: string } }> | undefined
  > {
    return await this.client.mutate<{ updateDatabaseSchema: { id: string } }>({
      mutation: UPDATE_DATABASESCHEMA_MUTATION,
      variables: {
        input: databaseSchema
      }
    });
  }

  public async deleteSchemas(
    ids: string[]
  ): Promise<
    | IDataAndErrors<{
        deleteDatabaseSchemas: IDeleteDatabaseSchemasResponse;
      }>
    | undefined
  > {
    return await this.client.mutate<{
      deleteDatabaseSchemas: IDeleteDatabaseSchemasResponse;
    }>({
      mutation: DELETE_DATABASESCHEMAS_MUTATION,
      variables: {
        input: {
          ids
        }
      }
    });
  }

  public async testJdbcConnectionForId(
    id: string
  ): Promise<
    | IDataAndErrors<{
        testJdbcConnectionForId: boolean;
      }>
    | undefined
  > {
    return await this.client.mutate<{ testJdbcConnectionForId: boolean }>({
      mutation: TEST_JDBC_CONNECTION_FOR_ID_MUTATION,
      variables: {
        id
      }
    });
  }

  public async testJdbcConnectionForJdbcUser(jdbcUser: IJdbcUser) {
    return await this.client.mutate<{
      testJdbcConnectionForJdbcUser: boolean;
    }>({
      mutation: TEST_JDBC_CONNECTION_FOR_JDBCUSER_MUTATION,
      variables: {
        input: jdbcUser
      }
    });
  }

  public async createDatabaseSchema(
    databaseSchema: ICreateDatabaseSchemaInput
  ) {
    return await this.client.mutate<{
      createDatabaseSchema: {
        id: string;
        jdbcUrl: string;
        users: [{ username: string; password: string }];
      };
    }>({
      mutation: CREATE_DATABASE_SCHEMA_MUTATION,
      variables: {
        input: databaseSchema
      }
    });
  }
}
