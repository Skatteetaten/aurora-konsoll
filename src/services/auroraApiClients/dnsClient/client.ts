import GoboClient, { IDataAndErrors } from 'services/GoboClient';
import { CnameInfosQuery, CNAME_INFO_QUERY } from './query';

export class DnsClient {
  private client: GoboClient;

  constructor(client: GoboClient) {
    this.client = client;
  }

  public async fetchCnameInfos(
    affiliation: string
  ): Promise<IDataAndErrors<CnameInfosQuery>> {
    return await this.client.query<CnameInfosQuery>({
      query: CNAME_INFO_QUERY,
      variables: {
        affiliation,
      },
    });
  }
}
