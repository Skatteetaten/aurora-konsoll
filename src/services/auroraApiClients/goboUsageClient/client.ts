import GoboClient, { IGoboResult } from 'services/GoboClient';
import { GOBO_USERS_USAGE_QUERY, IGoboUsage } from './query';

export class GoboUsageClient {
  private client: GoboClient;

  constructor(client: GoboClient) {
    this.client = client;
  }

  public async getGoboUsers(): Promise<IGoboResult<IGoboUsage> | undefined> {
    return await this.client.query<IGoboUsage>({
      query: GOBO_USERS_USAGE_QUERY
    });
  }
}
