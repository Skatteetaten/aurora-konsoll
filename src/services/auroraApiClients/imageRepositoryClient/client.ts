import ApolloClient from 'apollo-boost';

import { ITagsPagedGroup, TagsPagedGroup } from 'services/auroraApiClients';

import {
  IImageTagsConnection,
  ITagsGroupedQuery,
  ITagsQuery,
  TAGS_GROUPED_QUERY,
  TAGS_QUERY
} from './query';

export interface ITag {
  name: string;
  lastModified: string;
}

export interface ITagsPaged {
  endCursor: string;
  hasNextPage: boolean;
  tags: ITag[];
}

export interface IImageRepositoryClient {
  findTagsPaged: (
    repository: string,
    first?: number,
    cursor?: string,
    types?: string[]
  ) => Promise<ITagsPaged>;
  findGroupedTagsPaged: (repository: string) => Promise<TagsPagedGroup>;
}

export class ImageRepositoryClient implements IImageRepositoryClient {
  private client: ApolloClient<{}>;

  constructor(client: ApolloClient<{}>) {
    this.client = client;
  }

  public async findTagsPaged(
    repository: string,
    first?: number,
    cursor?: string,
    types?: string[]
  ): Promise<ITagsPaged> {
    const result = await this.client.query<ITagsQuery>({
      query: TAGS_QUERY,
      variables: {
        cursor,
        first,
        repositories: [repository],
        types
      }
    });

    const { imageRepositories } = result.data;

    if (!(imageRepositories && imageRepositories.length > 0)) {
      throw new Error(`Could not find tags for repository ${repository}`);
    }

    return this.toTagsPaged(imageRepositories[0].tags);
  }

  public async findGroupedTagsPaged(
    repository: string
  ): Promise<TagsPagedGroup> {
    const result = await this.client.query<ITagsGroupedQuery>({
      query: TAGS_GROUPED_QUERY,
      variables: {
        repositories: [repository]
      }
    });

    const { imageRepositories } = result.data;

    if (!(imageRepositories && imageRepositories.length > 0)) {
      throw new Error(`Could not find tags for repository ${repository}`);
    }

    const [mainRepo] = imageRepositories;

    const normalizedTags: ITagsPagedGroup = {
      auroraVersion: this.toTagsPaged(mainRepo.auroraVersion),
      bugfix: this.toTagsPaged(mainRepo.bugfix),
      latest: this.toTagsPaged(mainRepo.latest),
      major: this.toTagsPaged(mainRepo.major),
      minor: this.toTagsPaged(mainRepo.minor),
      snapshot: this.toTagsPaged(mainRepo.snapshot)
    };

    return new TagsPagedGroup(normalizedTags);
  }
  private toTagsPaged(imageTagsConnection: IImageTagsConnection): ITagsPaged {
    const { edges, pageInfo } = imageTagsConnection;
    return {
      endCursor: pageInfo.endCursor,
      hasNextPage: pageInfo.hasNextPage,
      tags: edges.map(edge => ({
        lastModified: edge.node.lastModified,
        name: edge.node.name
      }))
    };
  }
}
