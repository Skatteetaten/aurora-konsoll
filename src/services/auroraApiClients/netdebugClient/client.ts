import ApolloClient from 'apollo-boost';

import { IScanQuery, IScanStatusQuery, NETDEBUG_QUERY } from './query';

export interface INetdebugResult {
  status: string;
  open: IScanStatus[];
  failed: IScanStatus[];
}
export interface IScanStatus {
  status: string;
  message?: string;
  clusterNodeIp?: number;
}

export class NetdebugClient {
  private client: ApolloClient<{}>;

  constructor(client: ApolloClient<{}>) {
    this.client = client;
  }

  public async findNetdebugStatus(
    host: string,
    port: string
  ): Promise<INetdebugResult> {
    const result = await this.client.query<IScanQuery>({
      query: NETDEBUG_QUERY,
      variables: {
        host,
        port
      }
    });
    return this.showNetdebugStatus(result.data);
  }

  private showNetdebugStatus(item: IScanQuery): INetdebugResult {
    return {
      failed: this.normalizeScanStatus(item.scan.failed),
      open: this.normalizeScanStatus(item.scan.open),
      status: item.scan.status
    };
  }

  private normalizeScanStatus(scanStatus?: IScanStatusQuery): IScanStatus[] {
    if (!scanStatus) {
      return [];
    }

    return scanStatus.edges.map(edge => {
      const { clusterNode, message, status } = edge.node;
      return {
        clusterNodeIp: clusterNode && clusterNode.ip,
        message,
        status
      };
    });
  }
}
