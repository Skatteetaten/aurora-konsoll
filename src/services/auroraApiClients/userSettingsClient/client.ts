import GoboClient from 'services/GoboClient';
import { IUserSettings, USERSETTINGS_QUERY } from './query';

export class UserSettingsClient {
  private client: GoboClient;

  constructor(client: GoboClient) {
    this.client = client;
  }

  public async getUserSettings(): Promise<IUserSettings> {
    const result = await this.client.query<IUserSettings>({
      query: USERSETTINGS_QUERY
    });

    if (result && result.data) {
      return result.data;
    }

    return { userSettings: { applicationDeploymentFilters: [] } };
  }
}
