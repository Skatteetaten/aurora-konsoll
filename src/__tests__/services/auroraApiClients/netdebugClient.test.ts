import { NetdebugClient } from 'services/auroraApiClients/netdebugClient/client';
import { goboClientMock, GraphQLSeverMock } from 'utils/GraphQLMock';

import * as getNetdebugStatus from './__responses__/netdebugClient/getNetdebugStatus.json';
import * as getNetdebugStatusWhenScanIsNull from './__responses__/netdebugClient/getNetdebugStatusWhenScanIsNull.json';

const serverMock = new GraphQLSeverMock();
const clientMock = goboClientMock(serverMock.graphQLUrl);
const netdebugClient = new NetdebugClient(clientMock);

afterAll(() => {
  serverMock.close();
});

describe('getNetdebugStatus', () => {
  it('should fetch netdebug status from GraphQL server and normalize data', async () => {
    serverMock.putResponse('getNetdebugStatus', getNetdebugStatus);

    const result = await netdebugClient.findNetdebugStatus(
      'http://localhost',
      '8080'
    );
    expect(result).toMatchSnapshot();
  });

  it('should handle netdebug result when response is only null', async () => {
    serverMock.putResponse('getNetdebugStatus', null);

    const result = await netdebugClient.findNetdebugStatus(
      'http://localhost',
      '3000'
    );
    expect(result).toMatchSnapshot();
  });

  it('should handle netdebug result when scan is null', async () => {
    serverMock.putResponse(
      'getNetdebugStatus',
      getNetdebugStatusWhenScanIsNull
    );

    const result = await netdebugClient.findNetdebugStatus(
      'http://localhost',
      '4000'
    );
    expect(result).toMatchSnapshot();
  });
});
