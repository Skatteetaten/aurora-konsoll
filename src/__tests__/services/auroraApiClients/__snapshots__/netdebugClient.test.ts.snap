// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getNetdebugStatus should fetch netdebug status from GraphQL server and normalize data 1`] = `
Object {
  "data": Object {
    "scan": Object {
      "failed": Object {
        "edges": Array [
          Object {
            "node": Object {
              "clusterNode": Object {
                "ip": "127.0.0.1",
              },
              "message": "Klarte ikke resolve hostnavn til ip addresse.",
              "resolvedIp": "http://localhost",
              "status": "DNS_FAILED",
            },
          },
        ],
        "totalCount": 1,
      },
      "open": Object {
        "edges": Array [],
        "totalCount": 0,
      },
      "status": "CLOSED",
    },
  },
  "errors": undefined,
  "name": "getNetdebugStatus",
}
`;

exports[`getNetdebugStatus should handle netdebug result when response is only null 1`] = `undefined`;

exports[`getNetdebugStatus should handle netdebug result when scan is null 1`] = `
Object {
  "data": Object {
    "scan": null,
  },
  "errors": undefined,
  "name": "getNetdebugStatus",
}
`;
